tests/openedx_learning/core/components/test_api.py

```diff
@@ -0,0 +1,439 @@
+"""
+Basic tests of the Components API.
+"""
+from datetime import datetime, timezone
+
+from django.core.exceptions import ObjectDoesNotExist
+
+from openedx_learning.core.components import api as components_api
+from openedx_learning.core.contents import api as contents_api
+from openedx_learning.core.publishing import api as publishing_api
+from openedx_learning.lib.test_utils import TestCase
+
+
+class TestPerformance(TestCase):
+    """
+    Performance related tests to make sure we don't get n + 1 queries.
+    """
+    @classmethod
+    def setUpTestData(cls) -> None:
+        """
+        Initialize our content data (all our tests are read only).
+
+        We don't actually need to add content to the ComponentVersions, since
+        for this we only care about the metadata on Compnents, their versions,
+        and the associated draft/publish status.
+        """
+        cls.learning_package = publishing_api.create_learning_package(
+            "components.TestPerformance",
+            "Learning Package for Testing Performance (measured by DB queries)",
+        )
+        cls.now = datetime(2023, 5, 8, tzinfo=timezone.utc)
+
+    def test_component_num_queries(self) -> None:
+        """
+        Create a basic component and test that we fetch it back in 1 query.
+        """
+        component, _version = components_api.create_component_and_version(
+            learning_package_id=self.learning_package.id,
+            namespace="xblock.v1",
+            type="problem",
+            local_key="Query Counting",
+            title="Querying Counting Problem",
+            created=self.now,
+            created_by=None,
+        )
+        publishing_api.publish_all_drafts(
+            self.learning_package.pk,
+            published_at=self.now
+        )
+
+        # We should be fetching all of this with a select-related, so only one
+        # database query should happen here.
+        with self.assertNumQueries(1):
+            component = components_api.get_component(component.pk)
```

Comment:
Nit: the docstring of this class says it's "to make sure we don't get n + 1 queries", but calling `get_component` for a single component is not a situation in which you're going to see n + 1 queries. I guess you're still fleshing this out though. I'll look at the rest of the tests later when this is ready :)

Reply:
I meant that there aren't additional queries to get the draft and published versions of things, but yeah, that's not n+1. Will re-phrase.